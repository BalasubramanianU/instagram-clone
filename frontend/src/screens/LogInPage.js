import React, { useState, useEffect } from "react";
import "../css/styles.css";
import { Link } from "react-router-dom";
import {
  validateEmail,
  validateName,
  validateNumber,
  validatePassword,
} from "../utils/userValidation";

const LogInPage = () => {
  const [formData, setFormData] = useState({ logInId: "", password: "" });
  const [isValid, setIsValid] = useState(false);
  const [passwordField, setPasswordField] = useState({
    inputType: "password",
    button: "Show",
  });
  const [loginType, setLoginType] = useState("");
  const [error, setError] = useState({});

  useEffect(() => {
    if (formData.logInId.length > 0 && formData.password.length > 5) {
      return setIsValid(true);
    }
    setIsValid(false);
  }, [formData]);

  const handleChange = (event) => {
    setFormData({ ...formData, [event.target.name]: event.target.value });
  };

  const passwordButtonClick = () => {
    passwordField.button === "Show"
      ? setPasswordField({ inputType: "text", button: "Hide" })
      : setPasswordField({ inputType: "password", button: "Show" });
  };

  const handleSubmit = () => {
    // TODO: to be continued...
    if (validatePassword(formData.password).error) {
      setLoginType("");
      return setError({ password: true });
    }
    if (!validateEmail(formData.logInId).error) setLoginType("email");
    else if (!validateNumber(formData.logInId).error)
      setLoginType("mobileNumber");
    else if (!validateName(formData.logInId).error) setLoginType("userName");
    else {
      setLoginType("");
      return setError({ userName: true });
    }
    setError({});

    const requestOptions = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        [loginType]: formData.logInId,
        password: formData.password,
      }),
    };
    const apiCall = async () => {
      try {
        const response = await fetch(
          "http://192.168.1.8:5000/user/login",
          requestOptions
        );
        if (response.ok) {
          const data = await response.json();
          return console.log(data);
        }
        const errorMessage = await response.text();
        if (errorMessage.includes("User does not exist"))
          setError({ userName: true });
        if (errorMessage.includes("Invalid password"))
          setError({ password: true });
      } catch (error) {
        // all the errors in the try block will accumulate here, you need to
        // write logic here if you want to handle connection failed and other
        // exceptions/errors seperately.
        console.log(error);
      }
    };
    apiCall();
  };

  return (
    <div className="backgroundLayer">
      <div className="header"></div>
      <div className="gridContainer">
        <div className="welcomeImage">
          <div className="screenshot"></div>
        </div>
        <div className="loginBox">
          <div className="logo"></div>
          <div className="inputField">
            <input
              type="text"
              minLength="1"
              maxLength="255"
              placeholder="Phone number, username, or email"
              required
              name="logInId"
              value={formData.logInId}
              onChange={handleChange}
            />
            <span className="nameSpan">Phone number, username, or email</span>
          </div>
          <div className="inputField">
            <input
              type={passwordField.inputType}
              minLength="1"
              maxLength="1024"
              placeholder="Password"
              required
              name="password"
              value={formData.password}
              onChange={handleChange}
            />
            <span className="passwordSpan">Password</span>
            <button onClick={passwordButtonClick}>
              {passwordField.button}
            </button>
          </div>
          <div className="buttonContainer">
            <button
              className="button"
              onClick={handleSubmit}
              disabled={!isValid}
            >
              Log In
            </button>
          </div>
          <div className="orComponent">
            <div className="lineStrike"></div>
            <p className="orText">OR</p>
            <div className="lineStrike"></div>
          </div>
          <button className="loginButton">
            <div className="iconContainerLogin">
              <div className="fbIconLogin">
                <ion-icon name="logo-facebook"></ion-icon>
              </div>
            </div>
            Log in with Facebook
          </button>
          {error.userName && (
            <span className="errorMessage">
              The username you entered doesn't belong to an account. Please
              check your username and try again.
            </span>
          )}
          {error.password && (
            <span className="errorMessage">
              Sorry, your password was incorrect. Please double-check your
              password.
            </span>
          )}
          <button className="forgotPassword">
            <a className="forgotPasswordText" href>
              Forgot password?
            </a>
          </button>
        </div>
        <div className="switchAuthMethodBox">
          <p className="generalText">
            Don't have an account?
            <Link className="signUp" to="/signup">
              Sign up
            </Link>
          </p>
        </div>
        <div className="appStoreContainer">
          <p className="generalText">Get the app.</p>
          <div className="appStoreImages">
            <button className="googlePlay"></button>
            <button className="appStore"></button>
          </div>
        </div>
      </div>
      <div className="footerContainer">
        <div className="footer">
          <div className="footerFirstRow">
            <a className="footerText" href>
              About
            </a>
            <a className="footerText" href>
              Blog
            </a>
            <a className="footerText" href>
              Jobs
            </a>
            <a className="footerText" href>
              Help
            </a>
            <a className="footerText" href>
              API
            </a>
            <a className="footerText" href>
              Privacy
            </a>
            <a className="footerText" href>
              Terms
            </a>
            <a className="footerText" href>
              Top Accounts
            </a>
            <a className="footerText" href>
              Hashtags
            </a>
            <a className="footerText" href>
              Locations
            </a>
            <a className="footerText" href>
              Instagram Lite
            </a>
          </div>
          <div className="footerSecondRow">
            <select className="footerDropdown">
              <option>English</option>
            </select>
            <p className="smallText">&copy 2021 Instagram from Facebook</p>
          </div>
        </div>
      </div>
    </div>
  );
};

export default LogInPage;
