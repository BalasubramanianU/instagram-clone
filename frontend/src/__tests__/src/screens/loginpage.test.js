import { BrowserRouter } from "react-router-dom";
import { create, act } from "react-test-renderer";
import LogInPage from "../../../screens/LogInPage";

describe("input and button scenarios", () => {
  let tree, instance, element;
  beforeAll(() => {
    tree = create(
      <BrowserRouter>
        <LogInPage />
      </BrowserRouter>
    );
    instance = tree.root;
  });

  it("should render the login page correctly", () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should change the input field type if show button is clicked on password field", () => {
    element = instance.findByProps({ testId: "passwordButton" });
    element.props.onClick();
    element = instance.findByProps({ name: "password" });

    expect(element.props.type).toBe("text");
  });

  it("should change the input field type if hide button is clicked on password field", () => {
    element = instance.findByProps({ testId: "passwordButton" });
    element.props.onClick();
    element = instance.findByProps({ name: "password" });

    expect(element.props.type).toBe("password");
  });

  it("should disable login button if input values are empty in input field", () => {
    element = instance.findByProps({ testId: "loginButton" });

    expect(element.props.disabled).toBe(true);
  });

  it("should not disable login button if valid values are present in the input field", () => {
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "1234" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    element = instance.findByProps({ testId: "loginButton" });

    expect(element.props.disabled).toBe(false);
  });

  it("should display a error message when user logs in using invalid password", () => {
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "1234" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcd" } });
    });
    act(() => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });
    element = instance.findAllByType("span");
    const errorMsgLength = element.filter(
      (spanElement) =>
        spanElement.props.children ===
        "Sorry, your password was incorrect. Please double-check your password."
    ).length;

    expect(errorMsgLength).toBe(1);
  });

  it("should display a error message when user logs in using invalid user id", () => {
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    act(() => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });
    element = instance.findAllByType("span");
    const errorMsgLength = element.filter(
      (spanElement) =>
        spanElement.props.children ===
        "The username you entered doesn't belong to an account. Please check your username and try again."
    ).length;

    expect(errorMsgLength).toBe(1);
  });
});

describe("api call - positive scenarios", () => {
  let tree,
    instance,
    element,
    token = "json web token here";
  beforeAll(() => {
    tree = create(
      <BrowserRouter>
        <LogInPage />
      </BrowserRouter>
    );
    instance = tree.root;
  });

  afterEach(() => {
    localStorage.clear();
  });

  it("should receive a jwt token if login request is successful with mobileNumber", async () => {
    jest.spyOn(global, "fetch").mockResolvedValue({
      ok: true,
      headers: {
        get: jest.fn().mockReturnValue(token),
      },
      json: jest.fn().mockResolvedValue({ mobileNumber: "1234" }),
    });
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "1234" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    await act(async () => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });

    expect(localStorage.getItem("token")).toBe(token);
  });

  it("should receive a jwt token if login request is successful with email", async () => {
    jest.spyOn(global, "fetch").mockResolvedValue({
      ok: true,
      headers: {
        get: jest.fn().mockReturnValue(token),
      },
      json: jest.fn().mockResolvedValue({ email: "abc@abc.com" }),
    });
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({
        target: { name: "logInId", value: "abc@abc.com" },
      });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    await act(async () => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });

    expect(localStorage.getItem("token")).toBe(token);
  });

  it("should receive a jwt token if login request is successful with userName", async () => {
    jest.spyOn(global, "fetch").mockResolvedValue({
      ok: true,
      headers: {
        get: jest.fn().mockReturnValue(token),
      },
      json: jest.fn().mockResolvedValue({ userName: "abc" }),
    });
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "abc" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    await act(async () => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });

    expect(localStorage.getItem("token")).toBe(token);
  });
});

describe("api call - negative scenarios", () => {
  let tree, instance, element;
  beforeAll(() => {
    tree = create(
      <BrowserRouter>
        <LogInPage />
      </BrowserRouter>
    );
    instance = tree.root;
  });

  let mockResponse = (response) => {
    response
      ? jest.spyOn(global, "fetch").mockResolvedValue({
          text: jest.fn().mockResolvedValue(response),
        })
      : jest.spyOn(global, "fetch").mockRejectedValue("connection error");
  };

  it("should display a error message when user logs in using invalid user id", async () => {
    mockResponse("User does not exist");
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "1234" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    await act(async () => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });
    element = instance.findAllByType("span");
    const errorMsgLength = element.filter(
      (spanElement) =>
        spanElement.props.children ===
        "The username you entered doesn't belong to an account. Please check your username and try again."
    ).length;

    expect(errorMsgLength).toBe(1);
    expect(localStorage.getItem("token")).toBe(null);
  });

  it("should display a error message when user logs in using invalid password", async () => {
    mockResponse("Invalid password");
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "1234" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    await act(async () => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });
    element = instance.findAllByType("span");
    const errorMsgLength = element.filter(
      (spanElement) =>
        spanElement.props.children ===
        "Sorry, your password was incorrect. Please double-check your password."
    ).length;

    expect(errorMsgLength).toBe(1);
    expect(localStorage.getItem("token")).toBe(null);
  });

  it("should display the error message in console when api request fails", async () => {
    mockResponse();
    const consoleSpy = jest.spyOn(console, "log");
    act(() => {
      element = instance.findByProps({ name: "logInId" });
      element.props.onChange({ target: { name: "logInId", value: "1234" } });
    });
    act(() => {
      element = instance.findByProps({ name: "password" });
      element.props.onChange({ target: { name: "password", value: "abcdef" } });
    });
    await act(async () => {
      element = instance.findByProps({ testId: "loginButton" });
      element.props.onClick();
    });

    expect(consoleSpy).toHaveBeenCalledWith("connection error");
  });
});
