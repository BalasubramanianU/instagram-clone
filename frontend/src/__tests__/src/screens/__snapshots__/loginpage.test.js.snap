// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input and button scenarios should render the login page correctly 1`] = `
<div
  className="backgroundLayer"
>
  <div
    className="header"
  />
  <div
    className="gridContainer"
  >
    <div
      className="welcomeImage"
    >
      <div
        className="screenshot"
      />
    </div>
    <div
      className="loginBox"
    >
      <div
        className="logo"
      />
      <div
        className="inputField"
      >
        <input
          maxLength="255"
          minLength="1"
          name="logInId"
          onChange={[Function]}
          placeholder="Phone number, username, or email"
          required={true}
          type="text"
          value=""
        />
        <span
          className="nameSpan"
        >
          Phone number, username, or email
        </span>
      </div>
      <div
        className="inputField"
      >
        <input
          maxLength="1024"
          minLength="1"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
          value=""
        />
        <span
          className="passwordSpan"
        >
          Password
        </span>
        <button
          onClick={[Function]}
          testId="passwordButton"
        >
          Show
        </button>
      </div>
      <div
        className="buttonContainer"
      >
        <button
          className="button"
          disabled={true}
          onClick={[Function]}
          testId="loginButton"
        >
          Log In
        </button>
      </div>
      <div
        className="orComponent"
      >
        <div
          className="lineStrike"
        />
        <p
          className="orText"
        >
          OR
        </p>
        <div
          className="lineStrike"
        />
      </div>
      <button
        className="loginButton"
      >
        <div
          className="iconContainerLogin"
        >
          <div
            className="fbIconLogin"
          >
            <ion-icon
              name="logo-facebook"
            />
          </div>
        </div>
        Log in with Facebook
      </button>
      <button
        className="forgotPassword"
      >
        <a
          className="forgotPasswordText"
          href={true}
        >
          Forgot password?
        </a>
      </button>
    </div>
    <div
      className="switchAuthMethodBox"
    >
      <p
        className="generalText"
      >
        Don't have an account?
        <a
          className="signUp"
          href="/signup"
          onClick={[Function]}
        >
          Sign up
        </a>
      </p>
    </div>
    <div
      className="appStoreContainer"
    >
      <p
        className="generalText"
      >
        Get the app.
      </p>
      <div
        className="appStoreImages"
      >
        <button
          className="googlePlay"
        />
        <button
          className="appStore"
        />
      </div>
    </div>
  </div>
  <div
    className="footerContainer"
  >
    <div
      className="footer"
    >
      <div
        className="footerFirstRow"
      >
        <a
          className="footerText"
          href={true}
        >
          About
        </a>
        <a
          className="footerText"
          href={true}
        >
          Blog
        </a>
        <a
          className="footerText"
          href={true}
        >
          Jobs
        </a>
        <a
          className="footerText"
          href={true}
        >
          Help
        </a>
        <a
          className="footerText"
          href={true}
        >
          API
        </a>
        <a
          className="footerText"
          href={true}
        >
          Privacy
        </a>
        <a
          className="footerText"
          href={true}
        >
          Terms
        </a>
        <a
          className="footerText"
          href={true}
        >
          Top Accounts
        </a>
        <a
          className="footerText"
          href={true}
        >
          Hashtags
        </a>
        <a
          className="footerText"
          href={true}
        >
          Locations
        </a>
        <a
          className="footerText"
          href={true}
        >
          Instagram Lite
        </a>
      </div>
      <div
        className="footerSecondRow"
      >
        <select
          className="footerDropdown"
        >
          <option>
            English
          </option>
        </select>
        <p
          className="smallText"
        >
          &copy 2021 Instagram from Facebook
        </p>
      </div>
    </div>
  </div>
</div>
`;
